### ISSUE TITLE
Rule S2259 FP: symbolic execution does not enter the `for` loop

### ISSUE THREAD

**Issue by bzuu** [SonarSource/sonar-dotnet#890]
### Description

The following code snippet seems to trigger a false positive S2259 `'lastEx' is null on at least one execution path.` on the last line with the throw statement:

```csharp
public FileStream Blah()
{
	IOException lastEx = null;
	for (int i = 0; i < 10; i++)
	{
		try
		{
			return File.Open("somepath", FileMode.Open);
		}
		catch (IOException e)
		{
			lastEx = e;
		}
	}

	throw new InvalidOperationException(lastEx.Message);
}
```
When the for loop is removed S2259 is not raised on the last line.

### Expected behavior

No S2259 raised on the last line as either the code has returned or the lastEx value has been set.

### Actual behavior

Rule S2259 is raised both using the scanner for MSBuild as well as using SonarLint.

### Related information

* SonarC# Version 6.5 (build 3766)
* Visual Studio Enterprise 2017 Version 15.4.1
* Scanner for MSBuild Version 3.0.2
* SonarLint 3.6.0.2584

**Comment by sm-g** [comment]
same for `forest` here after update from 6.8.2 to 7.4.0:

```cs
DisjointForest<TModel> forest = null;
try
{
	forest = DisjointForest<TModel>.Create(models);
}
catch (ArgumentException ex) when (ex.Message.Contains("sss", StringComparison.OrdinalIgnoreCase))
{
	return result;
}

foreach (var item in Filter<TItem, TModel>(items))
{
	var count = forest.Leaves.Count();
}

```

**Comment by andrei-epure-sonarsource** [comment]
This is because our Symbolic Execution engine thinks the code will never enter the loop.
```
    class TestForLoop
    {
        string Foo()
        {
            string s = null;
            for (int i = 0; i < 10; i++)
            {
                s = "";
            }
            return s.Trim(); // Noncompliant FP due to integer constraints
        }
    }
```

**Comment by andrei-epure-sonarsource** [comment]
related to  #2528
